src/dropout.jl

```diff
@@ -0,0 +1,152 @@
+5×5 Matrix{Float64}:
+ 3.33333  3.33333  3.33333  3.33333  3.33333
+ 0.0      0.0      0.0      0.0      0.0
+ 0.0      0.0      0.0      0.0      0.0
+ 3.33333  3.33333  3.33333  3.33333  3.33333
+ 0.0      0.0      0.0      0.0      0.0
+
+julia> mean(dropout(ones(10^4, 5), 0.3, dims=1), dims=1)
+1×5 Matrix{Float64}:
+ 1.00571  1.00571  1.00571  1.00571  1.00571
+```
+"""
+dropout(A::AbstractArray, p::Real; dims = :) = dropout(_rng_from_array(A), A, p; dims)
+
+function dropout(rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(eltype(A))
+    0 <= p <= 1 || throw(ArgumentError("dropout expects a probability 0 <= p <= 1"))
+    if p > 0
+        dst = similar(A, T)
+        pT = convert(real(T), p)
+        _dropout!(rng, dst, A, pT, dims)
+    else
+        # Not so sure we want fast paths... this tries but doesn't guarantee type-stability,
+        # and the rrule does not have such a fast paths.
+        convert(AbstractArray{T}, A)
+    end
+end
+
+"""
+    dropout!(B, A, p; dims=:)
+
+This does exactly `B .= dropout(A, p; dims)`,
+or rather, it's the implementation of out-of-place [`dropout`](@ref).
+"""
+function dropout!(dst::AbstractArray, src::AbstractArray, p::Real; dims=:)
+    size(dst) == size(src) || throw(DimensionMismatch("dropout! expects output array the same size as input"))
+    0 <= p <= 1 || throw(ArgumentError("dropout expects a probability 0 <= p <= 1"))
+    if p > 0
+        rng = _rng_from_array(A)
+        pT = convert(real(eltype(dst)), p)
+        _dropout!(rng, dst, src, pT, dims)
+    else
+        copyto!(dst, src)
+    end
+end
+
+# This is the easy case in that we can safely use the output array for random numbers.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims::Colon)
+    val = convert(eltype(dst), 1/(1-p))
+    rand!(rng, dst)
+    # dst .= (dst.>p) .* val .* src  # hits a SIMD bug
+    _fast_broadcast!(dst, src) do q, x
+        (q>p) * val * x
+    end
+    dst
+end
+
+# For other dims, we we do need to allocate something.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims)
+    tmp = similar(dst, ntuple(d -> d in dims ? size(src,d) : 1, ndims(src)))
+    rand!(rng, tmp)
+    val = convert(eltype(dst), 1/(1-p))
+    # One-pass strategy:
+    # dst .= (tmp.>p) .* val .* src
+    # Two-pass strategy:
+    _fast_broadcast!(tmp) do q
+        (q>p) * val
+    end
+    dst .= tmp .* src
+end
+
+# The gradient needs to keep the random choices made, thus store at least a BitArray,
+# but the following way turns out to be faster & simpler:
+function ChainRulesCore.rrule(::typeof(dropout), rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(eltype(A))
+    val = convert(T, 1/(1-p))
+    keep = if dims isa Colon
+        similar(A, T) 
+    else
+        similar(A, T, ntuple(d -> d in dims ? size(A,d) : 1, ndims(A)))
+    end
+    rand!(rng, keep)
+    Y = @. (keep>p) * A * val
```

Comment:
If the mask idea doesn't work out, another idea for making this type stable while still allowing short-circuiting on 0 and 1 is to make a lazy mask wrapper which only creates the mask if necessary. Depending on how devious we want to be, we can even use `promote_op` or a custom `mask_type(xtype::Type{A}, ...) where A <: AbstractArray` to determine what type the wrapped mask should be (since we can't determine it in the short-circuiting case).